src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs

```diff
@@ -59,86 +55,88 @@ internal static class GlobalJsonReader
-        /// Locate a file in either the directory specified or a location in the
-        /// directory structure above that directory.
-        /// </summary>
-        private static string GetDirectoryNameOfFileAbove(string startingDirectory, string fileName)
-        {
-            // Canonicalize our starting location
-            var lookInDirectory = Path.GetFullPath(startingDirectory);
+            var lookInDirectory = startingDirectory;
 
             do
             {
-                // Construct the path that we will use to test against
-                var possibleFileDirectory = Path.Combine(lookInDirectory, fileName);
+                var possibleFile = new FileInfo(Path.Combine(lookInDirectory.FullName, fileName));
 
-                // If we successfully locate the file in the directory that we're
-                // looking in, simply return that location. Otherwise we'll
-                // keep moving up the tree.
-                if (File.Exists(possibleFileDirectory))
-                {
-                    // We've found the file, return the directory we found it in
-                    return lookInDirectory;
-                }
-                else
+                if (possibleFile.Exists)
                 {
-                    // GetDirectoryName will return null when we reach the root
-                    // terminating our search
-                    lookInDirectory = Path.GetDirectoryName(lookInDirectory);
+                    result = possibleFile.FullName;
+
+                    return true;
                 }
+
+                lookInDirectory = lookInDirectory.Parent;
             }
             while (lookInDirectory != null);
 
-            // When we didn't find the location, then return an empty string
-            return string.Empty;
+            return false;
         }
 
-        private static string NormalizePath(string path)
+        /// <summary>
+        /// Parses a global.json and returns the MSBuild SDK versions.
+        /// </summary>
+        /// <remarks>
+        /// NoInlining is enabled ensure that System.Runtime.Serialization.Json.dll isn't loaded unless the method is called.
+        /// </remarks>
+        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)]
+        private static Dictionary<string, string> ParseMSBuildSdksFromGlobalJson(string path)
         {
-            return FixFilePath(Path.GetFullPath(path));
+            Dictionary<string, string> sdks = null;
 
-        }
-        private static string FixFilePath(string path)
-        {
-            return string.IsNullOrEmpty(path) || Path.DirectorySeparatorChar == '\\' ? path : path.Replace('\\', '/');
+            using (var stream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read))
+            using (var reader = System.Runtime.Serialization.Json.JsonReaderWriterFactory.CreateJsonReader(stream, XmlDictionaryReaderQuotas.Max))
```

Comment:
Given NuGet is already targeting .NET Core 3.1 and later, is there a benefit to using System.Text.Json's JsonReader instead?

Reply:
For .NET Framework, we'd need to ship a separate assembly which I don't think would JIT as fast.  The size of `global.json` is very small and so the performance gain mainly comes from not having to load a big DLL to read it.  So I think `System.Text.Json` would probably not be a good candidate for this.  Ideally we'd use some native library to read it so its even faster.